In [1]:
import pandas as pd
import os
from sklearn.model_selection import train_test_split

C:\Users\kiran\anaconda3\Lib\site-packages\pandas\core\arrays\masked.py:60: UserWarning: Pandas requires version '1.3.6' or newer of 'bottleneck' (version '1.3.5' currently installed).
  from pandas.core import (


In [3]:
df = pd.read_excel("Exhibit+2.xlsx")
df

Cars  Sales \n(in 1,000 units)  Price\n(in lakh rupees)  \
0    Rocinante 1                   171.877                      6.1   
1    Rocinante 2                   139.796                      6.1   
2    Rocinante 3                   178.947                      9.9   
3    Rocinante 4                   140.022                      5.8   
4    Rocinante 5                   186.476                     10.0   
5    Rocinante 6                   192.123                      6.5   
6    Rocinante 7                   175.085                      5.5   
7    Rocinante 8                   146.882                      8.4   
8    Rocinante 9                   202.847                      6.6   
9   Rocinante 10                   149.933                      8.8   
10  Rocinante 11                   171.579                      9.5   
11  Rocinante 12                   201.512                      9.2   
12  Rocinante 13                   142.604                      8.1   
13  Rocinante 14                   135.158                      5.4   
14  Rocinante 15                   107.322                      5.6   
15  Rocinante 16                   199.699                      6.3   
16  Rocinante 17                   202.875                      5.8   
17  Rocinante 18                   204.406                      7.7   
18  Rocinante 19                   107.331                      7.2   
19  Rocinante 20                   190.345                      8.4   
20  Rocinante 21                   108.419                      8.4   
21  Rocinante 22                   150.249                      5.7   
22  Rocinante 23                   155.882                      5.5   
23  Rocinante 24                   100.980                      8.4   
24  Rocinante 25                   182.679                      5.6   
25  Rocinante 26                   166.752                      9.9   
26  Rocinante 27                   121.561                      6.2   
27  Rocinante 28                   174.256                      9.9   
28  Rocinante 29                   119.018                      6.5   
29  Rocinante 30                   169.842                      7.1   
30  Rocinante 31                   198.311                      9.6   
31  Rocinante 32                   204.875                      5.8   
32  Rocinante 33                   119.561                      6.2   
33  Rocinante 34                   203.875                      5.9   
34  Rocinante 35                   118.561                      6.1   

    Mileage\n(Km/ltr)  Top speed (Km/hr)  
0                15.8              168.2  
1                12.1              149.6  
2                17.0              173.4  
3                11.6              170.6  
4                17.2              175.0  
5                17.6              173.1  
6                16.0              184.6  
7                13.0              175.7  
8                19.3              166.7  
9                13.3              175.4  
10               15.8              150.9  
11               19.2              140.5  
12               11.7              150.2  
13               11.2              146.2  
14                7.8              151.2  
15               19.0              143.2  
16               19.7              178.9  
17               19.5              168.2  
18                7.5              164.7  
19               18.1              182.7  
20                7.9              165.7  
21               12.6              137.0  
22               12.9              173.9  
23                7.6              165.1  
24               16.7              160.1  
25               15.5              155.6  
26                9.5              155.4  
27               16.3              161.9  
28                9.1              146.8  
29               15.3              149.5  
30               19.1              146.5  
31               19.4              178.9  
32                9.4              1

In [4]:
list(df)

['Cars',
 'Sales \n(in 1,000 units)',
 'Price\n(in lakh rupees)',
 'Mileage\n(Km/ltr)',
 'Top speed (Km/hr)']

### Split the whole data

In [5]:
X = df[['Price\n(in lakh rupees)', 'Mileage\n(Km/ltr)', 'Top speed (Km/hr)']]    #independent variable
y =df['Sales \n(in 1,000 units)']                                               #dependent variable

### Split the dataset into train data and testing sets (80% train, 20% test)

In [6]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2)

In [7]:
X_train.shape, X_test.shape

((28, 3), (7, 3))

In [8]:
X_train

Price\n(in lakh rupees)  Mileage\n(Km/ltr)  Top speed (Km/hr)
5                       6.5               17.6              173.1
20                      8.4                7.9              165.7
11                      9.2               19.2              140.5
30                      9.6               19.1              146.5
29                      7.1               15.3              149.5
33                      5.9               19.4              175.9
28                      6.5                9.1              146.8
14                      5.6                7.8              151.2
7                       8.4               13.0              175.7
24                      5.6               16.7              160.1
18                      7.2                7.5              164.7
3                       5.8               11.6              170.6
31                      5.8               19.4              178.9
21                      5.7               12.6              137.0
23                      8.4                7.6              165.1
15                      6.3               19.0              143.2
12                      8.1               11.7              150.2
13                      5.4               11.2              146.2
0                       6.1               15.8              168.2
27                      9.9               16.3              161.9
4                      10.0               17.2              175.0
26                      6.2                9.5              155.4
10                      9.5               15.8              150.9
1                       6.1               12.1              149.6
34                      6.1                9.4              165.4
9                       8.8               13.3              175.4
2                       9.9               17.0              173.4
32                      6.2                9.4              175.4

In [9]:
X_test

Price\n(in lakh rupees)  Mileage\n(Km/ltr)  Top speed (Km/hr)
25                      9.9               15.5              155.6
19                      8.4               18.1              182.7
17                      7.7               19.5              168.2
22                      5.5               12.9              173.9
6                       5.5               16.0              184.6
8                       6.6               19.3              166.7
16                      5.8               19.7              178.9

In [10]:
len( X_train.shape)

2

In [11]:
len(y_train)

28

In [12]:
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error
import numpy as np

In [13]:
clf = LinearRegression()
clf.fit(X_train, y_train)

LinearRegression()

In [14]:
y_pred = clf.predict(X_test)
y_pred

array([169.0751481 , 191.3188099 , 203.78856103, 150.05102778,
       175.74945066, 202.91170303, 206.57323648])

### Calculation of Root Mean Squared Error

In [15]:
rmse = np.sqrt(mean_squared_error(y_test, y_pred))
print("Root Mean Squared Error (RMSE):", rmse)

Root Mean Squared Error (RMSE): 2.7992199143749597


####  An RMSE of 3.25 suggests that, on average, our model's predictions are about 3.25 units away from the actual values. A smaller RMSE indicates greater accuracy for the model.

### Predicting both training and testing data

In [16]:
y_train_pred = clf.predict(X_train)
y_test_pred = clf.predict(X_test)

In [17]:
train_rmse = np.sqrt(mean_squared_error(y_train, y_train_pred))
test_rmse = np.sqrt(mean_squared_error(y_test, y_test_pred))

In [18]:
print("Training RMSE:", train_rmse)
print("Testing RMSE:", test_rmse)

Training RMSE: 2.009029551604095
Testing RMSE: 2.7992199143749597


#### Training RMSE: 2.009 – Good fit on training data.
#### Testing RMSE: 2.799 – Slightly higher error on test data, indicating mild overfitting.
### Interpretation: The model performs well but could improve in generalization; minor tuning may help reduce the test error.

## Marengo32 Model

In [20]:
file_path = 'Exhibit+2.xlsx'
df = pd.read_excel(file_path, sheet_name='Marengo Models')
df

Cars  Sales \n(in 1,000 units)  Price\n(in lakh rupees)  \
0    Marengo 1                    20.896                     42.5   
1    Marengo 2                    31.048                     36.0   
2    Marengo 3                    29.904                     54.7   
3    Marengo 4                    28.792                     42.7   
4    Marengo 5                    16.776                     44.9   
5    Marengo 6                    18.928                     35.5   
6    Marengo 7                    22.776                     51.3   
7    Marengo 8                    36.824                     30.4   
8    Marengo 9                    22.216                     38.4   
9   Marengo 10                    35.456                     32.2   
10  Marengo 11                    28.576                     59.4   
11  Marengo 12                    33.648                     50.8   
12  Marengo 13                    33.440                     57.5   
13  Marengo 14                    34.912                     44.6   
14  Marengo 15                    24.016                     34.6   
15  Marengo 16                    20.992                     56.4   
16  Marengo 17                    21.696                     41.4   
17  Marengo 18                    17.832                     53.7   
18  Marengo 19                    25.624                     44.8   
19  Marengo 20                    21.920                     32.4   
20  Marengo 21                    24.792                     40.9   
21  Marengo 22                    22.256                     44.1   
22  Marengo 23                    15.488                     46.6   
23  Marengo 24                    21.640                     31.4   
24  Marengo 25                    23.264                     54.7   
25  Marengo 26                    19.328                     37.9   
26  Marengo 27                    27.216                     52.9   
27  Marengo 28                    23.096                     42.7   
28  Marengo 29                    17.824                     54.9   
29  Marengo 30                    19.144                     44.0   
30  Marengo 31                    21.600                     42.2   

    Mileage\n(Km/ltr)  Top speed (Km/hr)  
0                 9.3              199.4  
1                 9.7              235.2  
2                16.6              240.8  
3                11.7              232.5  
4                13.7              188.8  
5                 9.6              184.2  
6                13.7              207.7  
7                12.6              249.5  
8                16.2              175.8  
9                 9.6              245.6  
10               11.3              223.3  
11               13.6              240.0  
12               13.5              250.0  
13               11.1              247.4  
14               14.1              192.0  
15               13.4              201.2  
16               12.8              193.3  
17               16.0              202.0  
18               12.7              208.9  
19               13.6              176.9  
20               11.6              218.2  
21               11.4              190.0  
22               16.3              176.8  
23               14.5              178.9  
24               14.3              192.3  
25                9.4              184.1  
26               15.3              240.3  
27               15.0              192.1  
28               13.5              199.0  
29               11.8              181.1  
30                9.3              181.2

In [21]:
# Prepare the data
X_marengo  = df[['Price\n(in lakh rupees)', 'Mileage\n(Km/ltr)', 'Top speed (Km/hr)']]
y_marengo  = df['Sales \n(in 1,000 units)']

### Step-1: Split the data into training and testing sets (80% training, 20% testing)

In [22]:
X_train_mar, X_test_mar, y_train_mar, y_test_mar = train_test_split(X_marengo, y_marengo, test_size=0.2, random_state=42)

# Initialize and train a Linear Regression model
model_marengo = LinearRegression()
model_marengo.fit(X_train_mar, y_train_mar)

# Make predictions on the test set
y_pred_mar = model_marengo.predict(X_test_mar)

In [23]:
X_train_mar.shape, X_test_mar.shape

((24, 3), (7, 3))

### Step-2: Calculation of RMSE ( Root Mean Squared Error)

In [24]:
# Calculate RMSE for the Marengo model
rmse_mar = mean_squared_error(y_test_mar, y_pred_mar, squared=False)

# Print the RMSE result
print(f'Root Mean Squared Error(RMSE) for Marengo Model : {rmse_mar}')

Root Mean Squared Error(RMSE) for Marengo Model : 2.5385272221701207


### Result: The Root Mean Squared Error (RMSE) for the Marengo Model is 2.539.

#### Interpretation: An RMSE of 2.539 suggests that the model’s predictions are, on average, about 2.539 units away from the actual values. This indicates moderate accuracy; while the model captures some of the patterns in the data, there may still be room for improvement in precision. Reducing RMSE further could enhance predictive performance.

### Predicting both training and testing data

In [26]:
y_train_pred = model_marengo.predict(X_train_mar)
y_test_pred = model_marengo.predict(X_test_mar)

### Calculation of RMSE for both training and testing data

In [27]:
train_rmse_mar = mean_squared_error(y_train_mar, y_train_pred)
test_rmse_mar = mean_squared_error(y_test_mar, y_test_pred)

In [28]:
print("Training RMSE:", train_rmse_mar)
print("Testing RMSE:", test_rmse_mar)

Training RMSE: 4.499201616486006
Testing RMSE: 6.44412045769875


### Step-3: Determining whether the model is Overfitting or Underfitting

In [29]:
if train_rmse_mar < test_rmse_mar:
    print("The model may be overfitting.")
elif train_rmse_mar > test_rmse_mar:
    print("The model may be underfitting.")
else:
    print("The model is performing consistently on both training and testing data.")

The model may be overfitting.


### Interpretation
The model appears to be overfitting. Here's why:
The Training RMSE (4.50) is significantly lower than the Testing RMSE (6.44).
In a well-performing model, both Training RMSE and Testing RMSE should be close in value, indicating that the model generalizes well to new data.
Here, a lower Training RMSE means the model performs well on the training data but struggles with new data, indicated by the higher Testing RMSE.